File: TP1/TP01Q11 (✔)/PalindromoRec.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


//determina se string corresponde com a flag
bool stop(char str[], int tamanho) 
{
  bool stop = true;

  if (!(tamanho == 3 && str[0] == 'F' && str[1] == 'I' && str[2] == 'M')) {
    stop = false;
  }

  return stop;
}

//verifica se é pal ou não, caso não seja retorna false, caso seja retorna true
bool isPal(char *str, int i, int tamanho) 
{
    if (i >= tamanho / 2)
    {
        return true;
    }
    else if (str[i] != str[tamanho - i - 1])
    {
        return false;
    }
    else
    {
        return isPal(str, i + 1, tamanho);
    }
}

int main() 
{
  char str[1000];
  scanf(" %1000[^\n]", str);
  int tamanho = strlen(str);

  bool stopF = stop(str, tamanho);//chama a fun

  if (!stopF)//entra somente se nao corresponder a flag
  {
    bool pal = isPal(str, 0, tamanho);//chama a string para verificar enviando a string e o valor inicial de i

    if(pal == true)
    { 
        puts("SIM");
    }
    else
    {
        puts("NAO");
    }

    main();//chama novamente a main
  }

  return 0;
}

File: TP1/TP01Q10 (✔)/PalindromoRec.java
public class PalindromoRec {

    /*
     * verifica se a string é FIM para encerrar o programa
     */
    public static boolean stop(String str, int tamanho) {
        boolean stop = true;

        if (!(tamanho == 3 && str.charAt(0) == 'F' && str.charAt(1) == 'I' && str.charAt(2) == 'M')) {
            stop = false;
        }

        return stop;
    }


    /*
     * o metodo verifica se i é menor que tamanho, caso sim entra no condicional, depois verifica se é palindromo,
     * caso não for, iguala i com tamanho e seta a variavel como false, fazendo o loop nao repetir depois da proxima recursao e retornar
     * falso
     */
    public static boolean isPal(String str, int tamanho, int i, boolean pal)
    {
        if(i < tamanho)
        {
            if(!(str.charAt(i) == str.charAt(tamanho)))
            {
                pal = false;
                i = tamanho;
            }

            pal = isPal(str, --tamanho, ++i, pal);
        }

        return pal;
    }

    public static void codigoRec(boolean stop)
    {
        if (!stop) {
            String str = MyIO.readLine();
            int tamanho = str.length();
            stop = stop(str, tamanho);
            if(!stop)//só entra se a str for diferente de FIM
            {
                Boolean pal = isPal(str, tamanho - 1, 0, true); //puxa a função pra verificar pal
                if(pal) MyIO.println("SIM");
                else { MyIO.println("NAO");}
                codigoRec(stop);//puxa a função codigoRec novamente
            }
        }

        return;
    }

    public static void main(String[] args) {
        boolean stop = false;
        codigoRec(stop);//chama o metodo recursivo
    }

}